### PR TITLE
Update to 2.5.17 + change name of package to winflexbison_installer 

### PR DESCRIPTION
**Pull request by madebr**
- Updated to 2.5.17
- needs a testing/2.5.17 branch, so ci will fail
- changed name to winflexbison_installer because the recipe is binary only, so it can be used as `build_requires`
- ci: https://ci.appveyor.com/project/madebr/conan-winflexbison

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,7 +1,7 @@
 cmake_minimum_required(VERSION 3.0.0)
 project(cmake_wrapper)
 
-include(conanbuildinfo.cmake)
+include("${CMAKE_BINARY_DIR}/conanbuildinfo.cmake")
 conan_basic_setup()
 
-add_subdirectory("source_subfolder")
```
Comment (Croydon) [inline CMakeLists.txt:7]: This is common across all recipes, but don't rename this when it isn't for a very good reason

File: CMakeLists.txt [hunk 2]
```diff
@@ -1,7 +1,7 @@
 cmake_minimum_required(VERSION 3.0.0)
 project(cmake_wrapper)
 
-include(conanbuildinfo.cmake)
+include("${CMAKE_BINARY_DIR}/conanbuildinfo.cmake")
```
Comment (Croydon) [inline CMakeLists.txt:4]: Why this change?

File: CMakeLists.txt (hunk 1 above)
Comment (madebr) [reply CMakeLists.txt:7]: ok

File: conanfile.py [hunk 3]
```diff
@@ -17,38 +18,75 @@ class WinflexbisonConan(ConanFile):
...
-    settings = "os", "arch", "compiler", "build_type"
-    options = {"shared": [True, False]}
-    default_options = {"shared": False}
-    _source_subfolder = "source_subfolder"
+    settings = "arch", "os_build", "arch_build", "compiler", "build_type"
 
-    def config_options(self):
-        if self.settings.os != "Windows":
-            raise ConanInvalidConfiguration("winflexbison is only supported on Windows.")
+    _source_subfolder = "sources"
```
Comment (Croydon) [inline conanfile.py]: Undo please

File: CMakeLists.txt (hunk 2 above)
Comment (madebr) [reply CMakeLists.txt:4]: I use the package development flow when developing the recipes.
When building, this does not copy the sources from `self.source_folder` to `self.build_folder`,
so the include requires an explicit path.

File: conanfile.py [hunk 4]
```diff
@@ -1,14 +1,15 @@
...
 from conans import ConanFile, CMake, tools
 from conans.errors import ConanInvalidConfiguration
+import shutil
+import tempfile
 
 
 class WinflexbisonConan(ConanFile):
-    name = "winflexbison"
-    version = "2.5.16"
+    name = "winflexbison_installer"
```
Comment (Croydon) [inline conanfile.py]: Winflexbison is already in conan-center. We can't easily rename it now. See https://github.com/bincrafters/community/issues/489

File: conanfile.py (hunk 3 above)
Comment (madebr) [reply conanfile.py]: will do

File: conanfile.py (hunk 4 above)
Comment (madebr) [reply conanfile.py]: This is a newer version, so I don't think existing packages will break.

File: conanfile.py [hunk 5]
```diff
@@ -17,38 +18,75 @@ class WinflexbisonConan(ConanFile):
...
+        dlfilepath = os.path.join(tempfile.gettempdir(), filename)
+        if os.path.exists(dlfilepath) and not tools.get_env("WINFLEXBISON_FORCE_DOWNLOAD", False):
+            self.output.info("Skipping download. Using cached {}".format(dlfilepath))
+        else:
+            self.output.info("Downloading {} from {}".format(name, url))
+            tools.download(url, dlfilepath)
+        tools.check_sha256(dlfilepath, sha256)
+        tools.untargz(dlfilepath)
+
+        extracted_dir = "{}-{}".format(name, self.version)
```
Comment (Croydon) [inline conanfile.py]: This is way too much custom logic, please stay with the templates.

If you need to make modifications there should be a good reason.
If these are general reasons please make pull requests to the templates

File: conanfile.py [hunk 6]
```diff
@@ -17,38 +18,75 @@ class WinflexbisonConan(ConanFile):
...
 
     def build(self):
         cmake = CMake(self)
+        cmake.definitions["CMAKE_INSTALL_PREFIX"] = self._fake_package
         cmake.configure()
-        cmake.build()
+        cmake.build(args=["--config", str(self.settings.build_type)])
 
     def package(self):
-        self.copy(pattern="LICENSE", dst="licenses", src=self._source_subfolder)
```
Comment (Croydon) [inline conanfile.py:47]: What is about the LICENSE file for this recipe? It should be copied as well

File: conanfile.py [hunk 7]
```diff
@@ -17,38 +18,75 @@ class WinflexbisonConan(ConanFile):
...
+        with tools.chdir(self.build_folder):
+            try:
+                os.mkdir(self._fake_package)
+            except IOError:
+                pass
+            cmake.install(args=["--config", str(self.settings.build_type)])
+
+        self.copy(pattern="COPYING", dst="licenses", src=self._source_subfolder)
+
+        self.copy(pattern="*.exe", src=self._fake_package, dst="bin",keep_path=False)
```
Comment (Croydon) [inline conanfile.py:71]: What is `self._fake_package` about?

Again, please stay with the templates as close as possible. We have way too many recipes as custom logic would be easily maintainable.

File: conanfile.py (hunk 6 above)
Comment (madebr) [reply conanfile.py:47]: It is called `COPYING`

File: test_package/conanfile.py [hunk 8]
```diff
@@ -1,18 +1,14 @@
-#!/usr/bin/env python
```
Comment (Croydon) [inline test_package/conanfile.py:1]: templates...

File: conanfile.py (hunk 5 above)
Comment (madebr) [reply conanfile.py]: Ok, no more caching.

File: conanfile.py [hunk 9]
```diff
@@ -17,38 +18,75 @@ class WinflexbisonConan(ConanFile):
...
+        self.copy(pattern="*.h", src=self._fake_package, dst="include", keep_path=True)
+        self.copy(pattern="data/*", src=self._fake_package, dst="bin", keep_path=True)
+        shutil.copy(os.path.join(self.package_folder, "bin", "win_flex.exe"), os.path.join(self.package_folder, "bin", "flex.exe"))
+        shutil.copy(os.path.join(self.package_folder, "bin", "win_bison.exe"), os.path.join(self.package_folder, "bin", "bison.exe"))
 
     def package_info(self):
-        self.env_info.PATH.append(os.path.join(self.package_folder, "bin"))
+        bindir = os.path.join(self.package_folder, "bin")
```
Comment (Croydon) [inline conanfile.py:78]: templates...

File: conanfile.py (hunk 7 above)
Comment (madebr) [reply conanfile.py:71]: Instead of copying sources out of the build tree, I lest the package install itself.
Because the install procedure does use `bin`, `include`, `share` subdirectories,
I install to a `fake package`, from which the files are copied from in `package()`

File: test_package/conanfile.py (hunk 8 above)
Comment (madebr) [reply test_package/conanfile.py:1]: conanfile's are not scripts, so the templates should be modified.
Agreed?

**Review by Croydon — CHANGES_REQUESTED** [review]
* please stay with the templates as close as possible. Exceptions for good reasons are allowed but should be kept to a minimum. If you think that things can be better, make pull requests to our templates first.
  * we can't easily rename winflexbison to winflexbison_installer see https://github.com/bincrafters/community/issues/489. We would need to re-apply it to conan-center and deprecate the current one. I'm still not entirely sure we want to do that 🤔

File: conanfile.py (hunk 9 above)
Comment (madebr) [reply conanfile.py:78]: This is way more user friendly.
Templates should be changed.

File: conanfile.py (hunk 4 above)
Comment (madebr) [reply conanfile.py]: I will revert to winflexbison

File: conanfile.py (hunk 6 above)
Comment (Croydon) [reply conanfile.py:47]: Maybe for winflexbison itself, but not our recipe -> https://github.com/bincrafters/conan-winflexbison/commit/58da2e0afee11823199fec7d3fbee12e4cfabc09#diff-37854d19817c792316d481f5beb93cc7

All licenses of the packaged software + the license for the Conan recipe should be packaged

File: build.py [hunk 10]
```diff
@@ -2,10 +2,13 @@
 # -*- coding: utf-8 -*-
 
 
-from bincrafters import build_template_default
+from bincrafters import build_shared, build_template_default
 
 if __name__ == "__main__":
 
-    builder = build_template_default.get_builder()
+    builder = build_shared.get_builder()
```
Comment (Croydon) [inline build.py:9]: What are these changes about?

Comment (madebr) [reply build.py:9]: This only provides a binary, no libraries.
So adding builds for all combinations of build_type and compiler does not make sense.

Comment (madebr) [reply build.py:9]: For example: https://dl.bintray.com/bincrafters/public-conan/bincrafters/winflexbison/2.5.16/stable/package/3ca85915ce57a6607676849c47c2cee02c75209c/conanmanifest.txt

no `/lib`

File: conanfile.py (hunk 6 above)
Comment (madebr) [reply conanfile.py:47]: Indeed, it is not copied.
In my defence, this was also not done before. See [manifest](https://dl.bintray.com/bincrafters/public-conan/bincrafters/winflexbison/2.5.16/stable/package/3ca85915ce57a6607676849c47c2cee02c75209c/conanmanifest.txt)
I will add this.

Comment (madebr) [reply conanfile.py:47]: I've added it.

File: test_package/conanfile.py (hunk 8 above)
Comment (madebr) [reply test_package/conanfile.py:1]: https://github.com/bincrafters/conan-templates/pull/49

File: conanfile.py (hunk 9 above)
Comment (madebr) [reply conanfile.py:78]: https://github.com/bincrafters/conan-templates/pull/49

File: CMakeLists.txt (hunk 2 above)
Comment (madebr) [reply CMakeLists.txt:4]: https://github.com/bincrafters/conan-templates/pull/49

File: conanfile.py (hunk 7 above)
Comment (solvingj) [reply conanfile.py:71]: I don't understand, maybe i'm missing something.  If the project has a proper cmake install method defined, we should be able to just call `cmake.install()` , and the files should be copied directly to `package` sub-directories by default.  I'm trying to use my imagination, but can't see any advantage to having a `fake package` directory or function or anything. Are we eliminating a copy of some files, or working around some conflicting filenames or something?

File: build.py (hunk 10 above)
Comment (solvingj) [reply build.py:9]: Tbh, this is surprising that this package generates no libs.  Indeed if it only produces binaries, `_installer` might make sense.  But, as @Croydon said, it's accepted already, so lets publish under current name for now and discuss rename later.

File: conanfile.py (hunk 7 above)
Comment (madebr) [reply conanfile.py:71]: I use the `fake_package`  directory because the default `install` step does not copy the built files to the `bin` subdirectories.

File: build.py (hunk 10 above)
Comment (madebr) [reply build.py:9]: I reverted to the original name.
I also added the tests of flex and bison.
I think winflexbison does not need a library, because winflexbison adds the missing symbols to the generated sources.

File: conanfile.py (hunk 7 above)
Comment (madebr) [reply conanfile.py:71]: The [install rules in the cmake file install the files in the root package folder](
https://github.com/lexxmark/winflexbison/blob/a72f3d6b5102b65f064a5054ba634d3d62e94f41/CMakeLists.txt#L37-L47).
You can either `tools.replace_in_file` all these lines, or simply copy these files manually after it installed itself. I choose the latter because it is easier.

File: build.py (hunk 10 above)
Comment (madebr) [reply build.py:9]: Note that using flex for on visual studio, requires the `--wincompat` option.

File: CMakeLists.txt (hunk 2 above)
Comment (Croydon) [reply CMakeLists.txt:4]: Merged into the templates

File: test_package/conanfile.py (hunk 8 above)
Comment (Croydon) [reply test_package/conanfile.py:1]: Merged into the templates

File: conanfile.py (hunk 9 above)
Comment (Croydon) [reply conanfile.py:78]: Merged into the templates

**Comment by Croydon** [comment]
I have merged this partially via
6195f49e7d864fd07cc070c3c75cb4fd4a9ddf2c
027ed434425386247610961a66d8d7cb10d40765
30484963b5d469071969f71e587634af5eed0246
469d07f29b3e16d048cda10bbb2f3b3289b03565

If you still miss something could you please open a new pull request?

Particularly 469d07f29b3e16d048cda10bbb2f3b3289b03565 is a difference to your pull request, as it is stays closer to the templates which is really important in my opinion.

Thanks for your work!
